### PR TITLE
Document behaviour of ternary operator of literals.

### PR DESCRIPTION
**Pull request by nishant-sachdeva**
Helps #12677 

Clarifies the docs and adds tests.

### CODE REVIEW

File: libsolidity/analysis/TypeChecker.cpp [hunk 1]
```diff
@@ -1321,6 +1321,7 @@ bool TypeChecker::visit(VariableDeclarationStatement const& _statement)
 				valueComponentType->category() == Type::Category::RationalNumber &&
 				dynamic_cast<RationalNumberType const&>(*valueComponentType).isFractional() &&
 				valueComponentType->mobileType()
+				// Add Check here (?)
```
Comment (chriseth) [inline libsolidity/analysis/TypeChecker.cpp]: This is only used to generate a nice error message. What is actually happening here is that in line 1313, we check for an implicit conversion to the variable's type. This type is explicit and not somehow generated from the combination of other types, so this case is perfectly fine.

File: libsolidity/analysis/TypeChecker.cpp [hunk 2]
```diff
@@ -1364,6 +1364,8 @@ void TypeChecker::endVisit(ExpressionStatement const& _statement)
 {
 	if (type(_statement.expression())->category() == Type::Category::RationalNumber)
 		if (!dynamic_cast<RationalNumberType const&>(*type(_statement.expression())).mobileType())
+		// Add Check here (?)
+		// An inadvertent conversion could be happening in RationalType::mobileType()
 			m_errorReporter.typeError(3757_error, _statement.expression().location(), "Invalid rational number.");
```
Comment (cameel) [inline libsolidity/analysis/TypeChecker.cpp]: This checks literal expressions that are the only thing in a statement. I.e.

```soldidity
1 + 2 * 3;
```

Such statements have no effect so I think there's no point warning about the conversion here.

File: libsolidity/analysis/TypeChecker.cpp [hunk 3]
```diff
@@ -1411,6 +1413,10 @@ bool TypeChecker::visit(Conditional const& _conditional)
 	else if (trueType && falseType)
 	{
 		commonType = Type::commonType(trueType, falseType);
+		// Add Check here (?)
+		// Seems to be dealing with ternary operators.
+		// Our test case arose out of ternary operators. 
+		// An inadvertent conversion could be happening
```
Comment (cameel) [inline libsolidity/analysis/TypeChecker.cpp]: Yeah, I think this is the one we want.

File: libsolidity/analysis/TypeChecker.cpp [hunk 4]
```diff
@@ -1591,6 +1597,8 @@ bool TypeChecker::visit(TupleExpression const& _tuple)
 			// Note: code generation will visit each of the expression even if they are not assigned from.
 			if (types[i]->category() == Type::Category::RationalNumber && components.size() > 1)
 				if (!dynamic_cast<RationalNumberType const&>(*types[i]).mobileType())
+				// Add Check here (?)
+				// An inadvertent conversion could be happening in RationalType::mobileType()
 					m_errorReporter.fatalTypeError(3390_error, components[i]->location(), "Invalid rational number.");
```
Comment (cameel) [inline libsolidity/analysis/TypeChecker.cpp]: I think this one is for stuff like
```solidity
(uint32 a, uint64 b) = (1, 2);
```
or
```solidity
return (1, 2);
```

Currently you cannot use tuple components in expressions so I think there's no point warning about it. You can only assign/pass the whole tuple or use a destructuring assignment.

This will change though if we decide to implement #12635.

**Comment by cameel** [comment]
@nishant-sachdeva Looks like this check catches some cases where only one argument is a literal or where there already is a conversion:
- Perpetual Pools: `if (negativeY) return absoluteX > absoluteY ? -1 : int8 (1);`
- Euler: `return eTokenAddress == address(0) ? 0 : IERC20(token).balanceOf(eulerAddress);`
- Bleeps: `return eta == 1 ? 0 : eta;`
- Colony: `return unsatisfactory ? 0 : task.payouts[_role][_token];`

File: test/libsolidity/syntaxTests/literals/ternary_operator_return_type.sol [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+    function g() public returns (uint256 result)
+    {
+        flags[uint8(1)] = true;
+        flags[uint8(2)] = true;
+        flags[uint8(3)] = true;
+
+        result = (flags[0] ? uint256(100) : 0) +
+                    (flags[1] ? 110 : 0) +
+                    (flags[2] ? 125 : 0) +
+                    (flags[3] ? 150 : 0);
```
Comment (cameel) [inline test/libsolidity/syntaxTests/literals/ternary_operator_return_type.sol]:
```suggestion
        result =
            (flags[0] ? uint256(100) : 0) +
            (flags[1] ? 110 : 0) +
            (flags[2] ? 125 : 0) +
            (flags[3] ? 150 : 0);
```

File: libsolidity/analysis/TypeChecker.cpp [hunk 6]
```diff
@@ -1411,7 +1411,6 @@ bool TypeChecker::visit(Conditional const& _conditional)
 	else if (trueType && falseType)
 	{
 		commonType = Type::commonType(trueType, falseType);
-
```
Comment (cameel) [inline libsolidity/analysis/TypeChecker.cpp]: Looks like this PR is no longer modifying anything substantial in this file. Can you remove this change from the commit?

File: test/libsolidity/syntaxTests/literals/ternary_operator_return_type.sol [hunk 7]
```diff
@@ -0,0 +1,19 @@
+contract TestTernary
+{
+    mapping(uint8 => bool) flags;
+
+    function g() public returns (uint256 result)
+    {
+        flags[uint8(1)] = true;
+        flags[uint8(2)] = true;
+        flags[uint8(3)] = true;
```
Comment (cameel) [inline test/libsolidity/syntaxTests/literals/ternary_operator_return_type.sol]: Storage is not relevant to the test so I'd make it simpler:
```suggestion
        bool a = true;
        bool b = true;
        bool c = true;
```

File: test/libsolidity/syntaxTests/literals/ternary_operator_return_type.sol (hunk 5 above)
Comment (cameel) [inline test/libsolidity/syntaxTests/literals/ternary_operator_return_type.sol]: Let's include more cases None of them should produce a warning now but if we introduce it later, we'll want to see which cases it affects.

```solidity
        bool t = true;
        bool f = false;
        uint8 v255 = 255;
        uint8 v63 = 63;
        uint8 a;

        // Currently none of these should produce errors or warnings.
        // The result of the operator is always a limited-precision integer, even if all arguments are literals.

        a = (t ? 63 : 255) + (f ? 63 : 255);
        a = (t ? 0x3f : 0xff) + (f ? 0x3f : 0xff);
        a = (t ? uint8(63) : 255) + (f ? 63 : uint8(255));
        a = (t ? v63 : 255) + (f ? 63 : v255);

        a = (true ? 63 : 255) + (false ? 63 : 255);
        a = (true ? 0x3f : 0xff) + (false ? 0x3f : 0xff);
        a = (true ? uint8(63) : 255) + (false ? 63 : uint8(255));
        a = (true ? v63 : 255) + (false ? 63 : v255);

        a = (t ? 63 : 255) - (f ? 63 : 255);
        a = (t ? 63 : 255) * (f ? 63 : 255);
        a = (t ? 63 : 255) / (f ? 63 : 255);

        a = (t ? (true ? 63 : 255) : (false ? 63 : 255)) + (f ? (t ? 63 : 255) : (f ? 63 : 255));
        a = uint8(t ? 63 : 255) + uint8(f ? 63 : 255);
```

Also, let's rename the test to `ternary_operator_return_type_with_literal_arguments.sol`

Comment (cameel) [inline test/libsolidity/syntaxTests/literals/ternary_operator_return_type.sol]: I think we should also have a semantic test to verify the current unintuitive behavior. I.e. that these actually overflow:
```solidity
        uint16 a = (t ? 63 : 255) + (f ? 63 : 255);
```
```solidity
        uint16 a = (true ? 63 : 255) + (false ? 63 : 255);
```

File: docs/types/operators.rst [hunk 8]
```diff
@@ -15,6 +15,14 @@ and the type of the operator's result:
    operand, use the type of the right operand,
 3. otherwise, the operation is not allowed.
 
+.. Warning::
```
Comment (cameel) [inline docs/types/operators.rst]: The [section about literals](https://docs.soliditylang.org/en/latest/types.html#rational-and-integer-literals) needs clarification too.

First, it says `by using them together with a non-literal expression` and to me this does not imply that `true` and `false` count too. They are not "non-literal expressions".

Second, I think this section needs a warning too, pointing out the behavior of the ternary operator. I'd mention the case of array literals too (i.e. something like `[255][0]` being of `uint8` type and not a literal expression).

Comment (cameel) [inline docs/types/operators.rst]: Aren't these case-sensitive?
```suggestion
.. warning::
```

File: docs/types/operators.rst [hunk 9]
```diff
@@ -15,6 +15,14 @@ and the type of the operator's result:
...
+    The "type of the operand" is calculated through a function mobileType() that assigns, in case we're dealing with Rational Literals,
+    the largest possible type that is smaller than the size of the value. An unintended effect of this mechanism is that rational literals
+    ( that are not floats ) get converted to the largest possible integer type that just covers the size of the number.
+
+    For eg, a Rational literal 100 will get converted to uint8.
+    Changes like these can lead to overflow errors and other unintended consequences down the line during various operations.
```
Comment (cameel) [inline docs/types/operators.rst]: The C++ `mobileType()` function is an implementation detail so I don't think it should be mentioned in the docs.

Also, we do not have floats :) We do have fixed-point types but something like `1.23` really is just a rational literal and in our type system has the same type as integer literals. Such literals have unlimited precision so it's neither fixed- nor float-point. It's an exact representation.

I'd put that warning lower, either just below `If both are literal numbers, the operation is computed with arbitrary precision.` or at the end of the section. That paragraph already explains what a mobile type is.

And here's how I would reword it:

```rst
.. warning::
    The result of the ternary operator is never a literal, even if all of its operands are literals.
    The result type is the bigger of the mobile types of the integer operands.

    For example ``255 + (true ? 1 : 0)`` will revert due to arithmetic overflow.
    The reason is that ``(true ? 1 : 0)`` is of ``uint8`` type, which forces the addition to be performed in ``uint8`` as well,
    and 256 exceeds the range allowed for this type.

    Another consequence is that an expression like ``1.5 + 1.5`` is valid but ``1.5 + (true ? 1.5 : 2.5)`` is not.
    This is because the former is a literal expression evaluated in unlimited precision and only its final value matters.
    The latter involves a conversion of a fractional rational literal to an integer, which is currently disallowed.
```

**Review by cameel — CHANGES_REQUESTED** [review]

**Comment by cameel** [comment]
Please adjust the title of the PR and modify the description so that it no longer has `Closes #12677` - the PR no longer solves the issue, just clarifies the docs and adds tests.

File: docs/types/operators.rst [hunk 10]
```diff
@@ -26,6 +26,18 @@ except for comparison operators where the result is always ``bool``.
 The operators ``**`` (exponentiation), ``<<``  and ``>>`` use the type of the
 left operand for the operation and the result.
 
+.. warning::
```
Comment (chriseth) [inline docs/types/operators.rst]: Some suggestions:
 - please check how we call the operator in other parts of the documentation and at least also add `? ... : ...` as an alternative here, so that everyone knows what we are talking about
 - "literal" is more strict than what we want here. I think we use the term rational constant or something, please check.

At this point in the documentation, I think a "note" is better then a warning - we don't use a warning for `**` either.

File: docs/types/value-types.rst [hunk 11]
```diff
@@ -463,14 +463,22 @@ There is no additional semantic meaning added to a number literal containing und
 the underscores are ignored.
 
 Number literal expressions retain arbitrary precision until they are converted to a non-literal type (i.e. by
-using them together with a non-literal expression or by explicit conversion).
+using them together with a non-literal expression or with booleans like true/false or by explicit conversion).
```
Comment (chriseth) [inline docs/types/value-types.rst]:
```suggestion
using them together with anything else than a number literal expression (like boolean literals) or by explicit conversion).
```

File: docs/types/value-types.rst [hunk 12]
```diff
@@ -463,14 +463,22 @@ There is no additional semantic meaning added to a number literal containing und
...
 This means that computations do not overflow and divisions do not truncate
 in number literal expressions.
 
 For example, ``(2**800 + 1) - 2**800`` results in the constant ``1`` (of type ``uint8``)
 although intermediate results would not even fit the machine word size. Furthermore, ``.5 * 8`` results
 in the integer ``4`` (although non-integers were used in between).
 
+.. warning::
+    There are cases, such as those involving the ternary operator, where implicit conversions to types of a smaller size can happen.
```
Comment (chriseth) [inline docs/types/value-types.rst]: type smaller than what?

File: test/libsolidity/syntaxTests/literals/ternary_operator_return_type_with_literal_arguments.sol [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+        bool y = true;
+        bool z = true;
+        bool w = true;
+
+        result =
+            (x ? uint256(100) : 0) +
+            (y ? 110 : 0) +
+            (z ? 125 : 0) +
+            (w ? 150 : 0);
+
```
Comment (cameel) [inline test/libsolidity/syntaxTests/literals/ternary_operator_return_type_with_literal_arguments.sol]: I though you'd just replace this part with the stuff below. I think it's redundant now.

File: test/libsolidity/semanticTests/literals/ternary_operator_with_literal_types_overflow.sol [hunk 14]
```diff
@@ -0,0 +1,21 @@
+contract TestTernary
+{
+    function h() pure public returns (uint16 b)
+    {
+        b = (true ? 63 : 255) + (false ? 63 : 255);
+        return b;
```
Comment (cameel) [inline test/libsolidity/semanticTests/literals/ternary_operator_with_literal_types_overflow.sol]: You do not need to use `return` when you have a named return value. It's a bit redundant.

File: docs/types/operators.rst [hunk 15]
```diff
@@ -26,6 +26,18 @@ except for comparison operators where the result is always ``bool``.
 The operators ``**`` (exponentiation), ``<<``  and ``>>`` use the type of the
 left operand for the operation and the result.
 
+.. note::
+    The result of the ternary operator ``... ? ... : ...`` is never a rational, even if all of its operands are rationals.
```
Comment (cameel) [inline docs/types/operators.rst]: To be honest, the terminology around literals/constants is a bit unclear to me too. I'd say at least "rational constant" and I saw @chriseth suggest that. Do we actually refer to them as just "rationals"?
```suggestion
    The result of the ternary operator (``... ? ... : ...``) is never a rational constant, even if all of its operands are rational constants.
```

File: docs/types/operators.rst [hunk 16]
```diff
@@ -26,6 +26,18 @@ except for comparison operators where the result is always ``bool``.
...
+    The result type is the bigger of the mobile types of the integer operands.
```
Comment (cameel) [inline docs/types/operators.rst]: I originally suggested "integer", but maybe "numeric" would be more fitting here? It applies to fractional numbers too.
```suggestion
    The result type is the bigger of the mobile types of the numeric operands.
```

File: docs/types/operators.rst (hunk 15 above)
Comment (cameel) [reply docs/types/operators.rst]: Actually, after looking at the other file, maybe we should go with "number literal"/"number literal expression" for consistency?

File: docs/types/value-types.rst [hunk 17]
```diff
@@ -463,14 +463,23 @@ There is no additional semantic meaning added to a number literal containing und
...
 in the integer ``4`` (although non-integers were used in between).
 
+.. warning::
+    There are cases, such as those involving the ternary operator, where implicit conversions to sub-types of a smaller size can happen.
+    (For example, a rational number 100 will get converted to uint8 because uint8 is the smallest sized type that is sufficient to accommodate it)
+    This can lead to unintended overflows during calculations further down the pipeline.
+
+    For example ``255 + (true ? 1 : 0)`` will revert due to arithmetic overflow.
+    The reason is that ``(true ? 1 : 0)`` is of ``uint8`` type, which forces the addition to be performed in ``uint8`` as well,
+    and 256 exceeds the range allowed for this type.
```
Comment (cameel) [inline docs/types/value-types.rst]: Here's my take:
```suggestion
    While most operators produce a literal expression when applied to literals, there are certain operators that do not follow this pattern:
    - Ternary operator (``... ? ... : ...``),
    - Array subscript (``<array>[<index>]``).

    You might expect expressions like ``255 + (true ? 1 : 0)`` or ``255 + [1, 2, 3][0]`` to be equivalent to using the literal 256 directly, but in fact they are computed within the type ``uint8`` and can overflow.
```

File: docs/types/operators.rst (hunk 15 above)
Comment (nishant-sachdeva) [reply docs/types/operators.rst]: I'm pushing changes with "number literal" .
We can replace it further if we decide so

File: docs/types/operators.rst [hunk 18]
```diff
@@ -26,6 +26,18 @@ except for comparison operators where the result is always ``bool``.
 The operators ``**`` (exponentiation), ``<<``  and ``>>`` use the type of the
 left operand for the operation and the result.
 
+.. note::
+    The result of the ternary operator (``... ? ... : ...``) is never a number literal, even if all of its operands are number literals.
+    The result type is the bigger of the mobile types of the numeric operands.
```
Comment (chriseth) [inline docs/types/operators.rst]:
```suggestion
    The result type is determined from the types of the two operands in the same way as above, converting to their mobile type first if required.
```

Comment (chriseth) [reply docs/types/operators.rst]: The main reason to rephrase this is that "size" is difficult because of signed types.

File: docs/types/operators.rst [hunk 19]
```diff
@@ -26,6 +26,18 @@ except for comparison operators where the result is always ``bool``.
...
+
+    For example ``255 + (true ? 1 : 0)`` will revert due to arithmetic overflow.
```
Comment (chriseth) [inline docs/types/operators.rst]:
```suggestion
    As a consequence, ``255 + (true ? 1 : 0)`` will revert due to arithmetic overflow.
```

File: docs/types/operators.rst [hunk 20]
```diff
@@ -26,6 +26,18 @@ except for comparison operators where the result is always ``bool``.
 The operators ``**`` (exponentiation), ``<<``  and ``>>`` use the type of the
 left operand for the operation and the result.
 
+.. note::
```
Comment (chriseth) [inline docs/types/operators.rst]: I'm sorry - I just now opened this page in a wider context, and I think this should just be a section
```
Ternary Operator
--------------
```
instead of a note.

File: docs/types/operators.rst [hunk 21]
```diff
@@ -26,6 +26,19 @@ except for comparison operators where the result is always ``bool``.
 The operators ``**`` (exponentiation), ``<<``  and ``>>`` use the type of the
 left operand for the operation and the result.
 
+Ternary Operator
+----------------
+    The result of the ternary operator (``... ? ... : ...``) is never a number literal, even if all of its operands are number literals.
+    The result type is determined from the types of the two operands in the same way as above, converting to their mobile type first if required.
```
Comment (cameel) [inline docs/types/operators.rst]: This will probably not render properly if you indent the text like this. The header is enough to create a section, you don't have to indent the content in RST or Markdown. The indented text may be interpreted as a code block.

File: docs/types/value-types.rst [hunk 22]
```diff
@@ -463,14 +463,21 @@ There is no additional semantic meaning added to a number literal containing und
...
 in number literal expressions.
 
 For example, ``(2**800 + 1) - 2**800`` results in the constant ``1`` (of type ``uint8``)
 although intermediate results would not even fit the machine word size. Furthermore, ``.5 * 8`` results
 in the integer ``4`` (although non-integers were used in between).
 
+.. warning::
+    While most operators produce a literal expression when applied to literals, there are certain operators that do not follow this pattern:
+    - Ternary operator (``... ? ... : ...``),
+    - Array subscript (``<array>[<index>]``).
```
Comment (cameel) [inline docs/types/value-types.rst]: Looks like this does not render properly in RST. I think we need an empty newline before the list.
```suggestion
    While most operators produce a literal expression when applied to literals, there are certain operators that do not follow this pattern:

    - Ternary operator (``... ? ... : ...``),
    - Array subscript (``<array>[<index>]``).
```

**Review by cameel — COMMENTED** [review]
Looks like we're almost there but still needs minor syntax corrections to render properly in the docs :)

File: docs/types/operators.rst [hunk 23]
```diff
@@ -26,6 +26,19 @@ except for comparison operators where the result is always ``bool``.
 The operators ``**`` (exponentiation), ``<<``  and ``>>`` use the type of the
 left operand for the operation and the result.
 
+Ternary Operator
+----------------
+The result of the ternary operator (``... ? ... : ...``) is never a number literal, even if all of its operands are number literals.
```
Comment (chriseth) [inline docs/types/operators.rst]: Now that this is its own section, maybe we should start with a more generic introduction unrelated to number literals.

File: docs/types/operators.rst [hunk 24]
```diff
@@ -26,6 +26,24 @@ except for comparison operators where the result is always ``bool``.
 The operators ``**`` (exponentiation), ``<<``  and ``>>`` use the type of the
 left operand for the operation and the result.
 
+Ternary Operator
+----------------
+The ternary operator evaluates a given expression to a true or false value and then
+executes one of the two given statements depending upon the result of the evaluation.
```
Comment (ekpyron) [inline docs/types/operators.rst]: Technically, what you call ``statementA`` and ``statementB`` aren't statements, but also expressions...

File: docs/types/operators.rst [hunk 25]
```diff
@@ -26,6 +26,23 @@ except for comparison operators where the result is always ``bool``.
 The operators ``**`` (exponentiation), ``<<``  and ``>>`` use the type of the
 left operand for the operation and the result.
 
+Ternary Operator
+----------------
+The ternary operator is used in expressions of the form ``<expression> ? <trueExpression> : <falseExpression>``.
+It executes one of the two given expressions depending upon the result of the evaluation of the main ``expression``.
+If ``expression`` evaluates to True, then ``<trueExpression>`` will be executed, else, ``<falseExpression>`` is executed.
```
Comment (ekpyron) [inline docs/types/operators.rst]:
```suggestion
It evaluates one of the latter two given expressions depending upon the result of the evaluation of the main ``expression``.
If ``expression`` evaluates to ``true``, then ``<trueExpression>`` will be evaluated, otherwise ``<falseExpression>`` is evaluated.
```

File: docs/types/operators.rst [hunk 26]
```diff
@@ -26,6 +26,23 @@ except for comparison operators where the result is always ``bool``.
 The operators ``**`` (exponentiation), ``<<``  and ``>>`` use the type of the
 left operand for the operation and the result.
 
+Ternary Operator
+----------------
+The ternary operator is used in expressions of the form ``<expression> ? <trueExpression> : <falseExpression>``.
+It evaluates one of the latter two given expressions depending upon the result of the evaluation of the main ``expression``.
+If ``expression`` evaluates to ``true``, then ``<trueExpression>`` will be evaluated, otherwise ``<falseExpression>`` is evaluated.
```
Comment (ekpyron) [inline docs/types/operators.rst]:
```suggestion
It evaluates one of the latter two given expressions depending upon the result of the evaluation of the main ``<expression>``.
If ``<expression>`` evaluates to ``true``, then ``<trueExpression>`` will be evaluated, otherwise ``<falseExpression>`` is evaluated.
```
Otherwise it's inconsistend with the use of ``<trueExpression>`` and ``<falseExpression>`` isn't it?

File: docs/types/operators.rst [hunk 27]
```diff
@@ -26,6 +26,23 @@ except for comparison operators where the result is always ``bool``.
 The operators ``**`` (exponentiation), ``<<``  and ``>>`` use the type of the
 left operand for the operation and the result.
 
+Ternary Operator
+----------------
+The ternary operator is used in expressions of the form ``<expression> ? <trueExpression> : <falseExpression>``.
+It evaluates one of the latter two given expressions depending upon the result of the evaluation of the main ``<expression>``.
+If ``<expression>`` evaluates to ``true``, then ``<trueExpression>`` will be evaluated, otherwise ``<falseExpression>`` is evaluated.
+
+The result of the ternary operator (``... ? ... : ...``) is never a number literal, even if all of its operands are number literals.
```
Comment (ekpyron) [inline docs/types/operators.rst]:
```suggestion
The result of the ternary operator is never a number literal, even if all of its operands are number literals.
```
We don't need to remention the form of ternary expressions now that we started off with it.

But I'm also not sure if we should say that the result "is never a number literal"... That's a bit of a weird thing to say: a complex expression obviously isn't a literal... technically, it would be accurate to say "never has rational number type", but I'm not sure that really makes things clear, so I'm not sure what's best to do here...

File: docs/types/operators.rst [hunk 28]
```diff
@@ -26,6 +26,23 @@ except for comparison operators where the result is always ``bool``.
 The operators ``**`` (exponentiation), ``<<``  and ``>>`` use the type of the
 left operand for the operation and the result.
 
+Ternary Operator
+----------------
+The ternary operator is used in expressions of the form ``<expression> ? <trueExpression> : <falseExpression>``.
+It evaluates one of the latter two given expressions depending upon the result of the evaluation of the main ``<expression>``.
+If ``<expression>`` evaluates to ``true``, then ``<trueExpression>`` will be evaluated, otherwise ``<falseExpression>`` is evaluated.
+
+The result of the ternary operator cannot have a rational number type, even if all of its operands are rational number literals.
```
Comment (ekpyron) [inline docs/types/operators.rst]:
```suggestion
The result of the ternary operator does not have a rational number type, even if all of its operands are rational number literals.
```
:-)?

**Review by Aivolt1 — APPROVED** [review]

**Review by cameel — DISMISSED** [review]
I think it's good enough in the current form. All the problems that were pointed out seem to have been fixed and I don't see any new ones so let's get it merged :)

Some tests are failing but they all look unrelated to the PR.

**Review by chriseth — APPROVED** [review]
